### PR TITLE
DBZ-988 Test MongoDB Unwrapper for diverse update ops

### PR DESCRIPTION
**Pull request by renatomefi**
https://issues.jboss.org/browse/DBZ-988

The idea of this PR is to improve the test coverage in the integration tests, with the intent to:
- Be aware of MongoDB BC breaks
- Build confidence when changing the Unwrap functionality
- Cover as many operations as possible
- In case of need split tests/functionalities by MongoDB version

ToDo:
- [x] Basic test setup
- [x] Test Field Operators https://docs.mongodb.com/v3.6/reference/operator/update/#fields
  - [x] $inc
  - [x] $min
  - [x] $max
  - [x] $mul
  - [x] $rename (Depends on DBZ-612)
  - [x] $set (Already tested multiple times in the current tests, basically the only operation tested)
  - [x] $setOnInsert
  - [x] $unset (Depends on DBZ-612)
- [x] Test Array Operators https://docs.mongodb.com/v3.6/reference/operator/update/#array
  - [x] $addToSet
  - [x] $pop
  - [x] $pull
  - [x] $push
  - [x] $pullAll
  - `$(update)` - Skipped since it's Mongo 3.6+
  - `$[]` - Skipped since it's Mongo 3.6+
  - `$[<identifier>]` - Skipped since it's Mongo 3.6+
  - Test modifiers https://docs.mongodb.com/v3.6/reference/operator/update/#modifiers - Skipped due to only changing the position of the affected items in the array
- [x] Test bitwise https://docs.mongodb.com/v3.6/reference/operator/update/#bitwise
  - [x] And
  - [x] Or
  - [x] Xor
- [x] Include mongo 3.2, 3.4, 3.6 and 4.0 into jenkins jobs (https://github.com/debezium/debezium/pull/677)

Before I continue with many hours of work I'd like to hear some feedback and whether it's desired in this way, if not I'm happy to learn and implement it differently!

### CODE REVIEW

**Comment by gunnarmorling** [comment]
Hey @renatomefi, it's definitely an excellent idea to improve testing for that SMT. As you say, there's lot to improve, which is one of the reasons why the SMT still is marked as "experimental" currently. That said, I did some checking when the SMT was originally contributed and it seemed most (all?) of the ops would be represented as simple set/unset ones in the logs we receive.

There could be a JIRA issue for it already; can you take a look at the open issues for the "mongodb-connector" component and see whether you find it? If not, can you file a new one and refer to it in the commit message?

I'll take a look at your approach tomorrow; one thing that's not ideal about the current tests is that they assemble the document to SMT runs against themselves. It'd be better to receive the actual change documents from the database in order to guard against regressions. 

Thanks a lot for your contributions to the SMT and the connector, that's much appreciated. The SMT recently didn't receive as much attention as it should have (just due to lack of time on our end), so it's great to see you stepping in end helping out 👍

File: debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelopeOperationsTestIT.java [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+
+import static org.fest.assertions.Assertions.assertThat;
+import static org.junit.Assert.fail;
+
+/**
+ * Integration test for {@link UnwrapFromMongoDbEnvelope}. It sends operations into
+ * MongoDB and listens on messages that are generated by Debezium plug-in. The messages
+ * are then run through the SMT itself.
+ * <p>
+ * This tries to cover evey mongo update operation as described in the official documentation
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelopeOperationsTestIT.java]:
```suggestion
 * This tries to cover every mongo update operation as described in the official documentation
```

File: debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelopeOperationsTestIT.java [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+        // Start the connector ...
+        start(MongoDbConnector.class, config);
+    }
+
+    @After
+    public void afterEach() {
+        try {
+            stopConnector();
+        } finally {
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelopeOperationsTestIT.java]: Can you please move the `finally` to the next line? The project style is to have no code after closing braces (it's not like that everywhere yet, but we'd like to converge to that eventually).

File: debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelopeOperationsTestIT.java [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+        records = consumeRecordsByTopic(1);
+        final SourceRecord candidateRecord = records.recordsForTopic(TOPIC_NAME).get(0);
+
+        if (((Struct) candidateRecord.value()).get("op").equals(Envelope.Operation.CREATE)) {
+            // MongoDB is not providing really consistent snapshot, so the initial insert
+            // can arrive both in initial sync snapshot and in oplog
+            records = consumeRecordsByTopic(1);
+        }
+
+        assertThat(records.recordsForTopic(TOPIC_NAME).size()).isEqualTo(1);
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelopeOperationsTestIT.java]: You can do `assertThat(xyz).hasSize(1)`.

File: debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelopeOperationsTestIT.java [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    @Ignore("Depends on https://issues.jboss.org/browse/DBZ-612")
+    public void shouldTransformOperationRename() throws InterruptedException {
+        SourceRecord updateRecord = executeSimpleUpdateOperation(
+                "{'$rename': {'dataInt': 'dataIntNewName', 'nonExistentField': 'nonExistentFieldRenamed'}}"
+        );
+
+        final SourceRecord transformedUpdate = transformation.apply(updateRecord);
+        final Struct transformedUpdateValue = (Struct) transformedUpdate.value();
+
+        assertThat(transformedUpdate.valueSchema().field("id").schema()).isEqualTo(Schema.OPTIONAL_INT32_SCHEMA);
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelopeOperationsTestIT.java]: How about fully assembling the expected value schema and then asserting on that (see `VerifyRecord#assertConnectSchemasAreEqual()`)?

**Comment by renatomefi** [comment]
@gunnarmorling thanks for the review!

Regarding:
> it seemed most (all?) of the ops would be represented as simple set/unset ones in the logs we receive

You're correct and that's something I noticed while working with the mongo oplog as well, in this case those tests will be present only to guarantee future compatibility, but if you want to I don't have to include them!
For the important part is to test operations which can generate
- Only a `$set`
- Only a `$unset`
- Both `$set` and `$unset`
Also ofc the array operations and nested fields, since we want to guarantee our unwrapper won't modify the final encoded json!

Should I proceed with all those tests or should I keep only the most problematic ones?

File: debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelopeOperationsTestIT.java (hunk 4 above)
Comment (renatomefi) [reply debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelopeOperationsTestIT.java]: It's great! Should I just use or move to a more "generic place"? I'm afraid of crossing packages boundaries here

**Comment by renatomefi** [comment]
Closed in favor of #682

**Comment by renatomefi** [comment]
Closen by mistake, this is still WIP and will later be rebased when #682 is merged

**Comment by gunnarmorling** [comment]
Hey @renatomefi, I think this PR is the one you're still planning to finsish, as you mentioned the other day in the chat, right?

**Comment by renatomefi** [comment]
hey @gunnarmorling, indeed, I'll refactor a bit the basic test to use the VerifyRecord and then I'm flooding it with tests :)

**Comment by gunnarmorling** [comment]
Hey @renatomefi, seems you just pushed an update? Thanks a lot; we're in holiday shutdown right now and will get back to review this once we're back from the holidays.

Thanks a ton for all your contributions in 2018 and and looking forward to collaborating with you in 2019, too! 🎉

**Comment by renatomefi** [comment]
Hello @gunnarmorling!

Yes, this is ready to review now, I did cover all Update operations compatible with MongoDB 3.2, maybe in the future I can add the 3.4 and 3.6 ones but I have to filter the tests!

Thanks you too, it was really great and it has been a great experience for me!

**Comment by gunnarmorling** [comment]
@renatomefi, I'm seeing this failure:

    UnwrapFromMongoDbEnvelopeTestIT.shouldDropTombstoneByDefault:116 expected:<[2]> but was:<[1]>

Does this one pass for you?

**Comment by gunnarmorling** [comment]
@renatomefi Looks great overall. Only curious whether you also saw that test failure above (perhaps intermittendendly)? And could you prefix the message of the last commit with "*D*BZ-..."? Thanks!

**Comment by renatomefi** [comment]
Hello @gunnarmorling, I did see it once, but when I test the whole suite with mvn it does happen. How did you do it?

**Comment by gunnarmorling** [comment]
I just ran this:

    mvn clean install -pl debezium-connector-mongodb

>

**Comment by renatomefi** [comment]
@gunnarmorling It's indeed an intermittent issue, the new tests have a way to "discard" the invalid update messages coming from the oplog when an insert happens (More details here: ), so I thought it was a better approach just to migrate `UnwrapFromMongoDbEnvelopeTestIT` to the new TestIT abstraction which helpers taking care of it!
I've added 2 new commits, you can review those and try to run mvn locally again to see if it's stable now :)

Thanks

**Comment by gunnarmorling** [comment]
@renatomefi Yes, it's looking better now. Going to merge it; thanks a lot!

**Pull request merged**